### ISSUE TITLE
Calling copy-config RPC from restconf

### ISSUE THREAD

**Issue by shmuelhazan** [clicon/clixon#158]
Hi, 

I am trying to run the copy-config RPC from restconf but I am getting strange errors:
```
$ curl -Ssik -X POST -H "Content-Type: application/yang-data+json" -d '{"ietf-netconf:input": {"target": {"startup": [null]},"source": {"running": [null]}}}' https://127.0.0.1/restconf/operations/ietf-netconf:copy-config
HTTP/2 400 
server: nginx/1.17.0
date: Mon, 16 Nov 2020 20:08:58 GMT
content_type: application/yang-data+json

{
"ietf-restconf:errors" : {
  "error": {
    "error-type": "rpc",
    "error-tag": "malformed-message",
    "error-severity": "error",
    "error-message": "restconf RPC does not have single input"
  }
}

}
```

With xml:
```
$ curl -H "Content-Type: application/yang-data+xml" https://10.40.3.126/restconf/operations/ietf-netconf:copy-config --insecure -X POST -d "<input xmlns=\"urn:ietf:params:xml:ns:netconf:base:1.0\"><source><running/></source><target><startup/></target></input>" -Ssik
HTTP/2 400 
server: nginx/1.17.0
date: Mon, 16 Nov 2020 20:31:01 GMT
content_type: application/yang-data+json

{
"ietf-restconf:errors" : {
  "error": {
    "error-type": "rpc",
    "error-tag": "malformed-message",
    "error-severity": "error",
    "error-message": "restconf RPC does not have single input"
  }
}

}
```

Am I missing something?

**Comment by dima1308** [comment]
Is it possible to do some workaround (in CLIXON or in caller application) to overcome the failure?

**Comment by olofhagsand** [comment]
I have difficulty repeating the error. I made some minor changes and made a patch  https://github.com/clicon/clixon/commit/a71e8aacebd93122e1736e857b5162121455f139 moments ago with a testcase: https://github.com/clicon/clixon/blob/a71e8aacebd93122e1736e857b5162121455f139/test/test_copy_config.sh#L169 for this.
Can you please upgrade and test again, and possibly try the testcase.
If it still does not work:
- are you running restconf nginx / fcgi or evhtp?
- can you run clixon_restconf -D 1 and observe syslog?
- The error message is at https://github.com/clicon/clixon/blob/a71e8aacebd93122e1736e857b5162121455f139/apps/restconf/restconf_methods_post.c#L619 it would be interesting to make a xml_print(stderr, xret) in a debugger at that point or in-compiled to see how the request looks like.

**Comment by dima1308** [comment]
it is problematic for me to execute a remote debugger (we use CLIXON as part of the embedded system). I will try the lastest CLIXON master and will update you

**Comment by dima1308** [comment]
1. I'm using NGNIX / fcgi
2. I used the latest CLIXON master - the problem is still.
3. please see attached log
[restconf.log](https://github.com/clicon/clixon/files/5643853/restconf.log)

**Comment by olofhagsand** [comment]
Debugging found that an extra reply from the backend plugin triggered a double reply that in turn triggered an invalid error check. 
Since it was an extra application callback it was not repeatable by system regression.
According to the log this is what was returned from the backend:
```
  <rpc-reply xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">
     <ok/>
  </rpc-reply>
  <rpc-reply/>  <------- Extra reply triggered by application?
```
The invalid error check has been corrected (see patch above) but please also check the extra callback of copy-config.

**Comment by dima1308** [comment]
@olofhagsand thanks, now it is working. We will do more tests tomorrow

**Comment by olofhagsand** [comment]
Actually fix: https://github.com/clicon/clixon/commit/3c014184f7e83a6f986a6568b14f0357771d882b was wrong. New patch https://github.com/clicon/clixon/commit/344dcacf6fbb36c5c8cfe5e6af36c924ad914d84 removes the check.
The main problem is actually backend plugin code that makes two replies to the same copy-config rpc.
But it should be detected in the backend, not in the restconf code.
@dima1308 please address this problem in your backend plugin, I will make a sanity check later to detect such behaviour.

**Comment by shmuelhazan** [comment]
@olofhagsand Thanks for the fix!

Our plugin indeed registers a callback for the copy-config RPCS (and fills the cbret value with a reply). What is the expected behavior from a plugin in this case?

**Comment by Topitzik** [comment]
Also, last time we talked about it, you suggested checking whether the previous response is still in cbret (assuming not an error) and if so editing it. unfortunately, it is empty when we get it (at least not for copy-config).

**Comment by olofhagsand** [comment]
This is a little tricky.
First it has to do which order the callbacks are registered.
The order is:
1. plugin_init
2. backend_rpc_init
3. plugin_start
Which means if you register a copy-config callback in (1), it will be called _before_ the system copy-config callback registered from (2) backend_rpc_init. If you register a copy-config in (3) plugin-start it will be called _after_ the system copy-config.

Second, if there are more than one reply (eg <rpc-reply/><rpc-reply/>) only the first reply will be parsed and used by the cli/netconf/restconf clients.

If you want to take the original and modify it, you should therefore register the callback in plugin_start so that your callback will be called second. Then you should modify the original reply (not add a new reply).

**Comment by Topitzik** [comment]
actually we register our RPCs during (3) but still get an empty cbret for copy-config.

**Comment by olofhagsand** [comment]
OK. That seems different from my analysis. The cbret should contain the previous.
If you see the printout (from you?):
```
<rpc-reply xmlns="urn:ietf:params:xml:ns:netconf:base:1.0">  <--- System copy-config
     <ok/>
  </rpc-reply>
  <rpc-reply/>  <--- application copy-config
```
Which follows that you register according to (3).
But in that case, you append the reply to the existing and you should have the system copy-config in the cbret?

**Comment by Topitzik** [comment]
ohh, we DO get the cbret with the previous reply. 
I had a logging error, sorry.

so modifying the provided cbret should be regarded as the right way for any following RPC?
since this does not happen by the backend, if we do decide to register an RPC at (1), how can we return any non-error reply without the backend itself adding the "extra" reply?

**Comment by olofhagsand** [comment]
This is somewhat beyond advanced usage - it is not really designed, but I guess that would work.
I see other people as well using multiple callbacks for different reasons.
One would probably need to make a more structured design to catch a more formal API.

**Comment by Topitzik** [comment]
so to summarise the current behavior and possible use cases:
1. we can only do RPCs registered at (3) by modifying cbret (as discussed above).
2. we can call custom RPCs (i.e. not netconf defined, like `copy-config` ) registered at (1) if we create the `rpc-reply` / `rpc-error` .
3. the only use case of an RPC which is also defined by netconf protocol and registered at (1) is to basically "forbid" an RPC completely by always returning an error (BE can only handle one of `rpc-reply` or `rpc-error` but `rpc-reply` will cause problem by the next RPC).

is that correct?

**Comment by olofhagsand** [comment]
I would rephrase it slightly:
1. The main usecase is user-defined RPC:s ie yang-defined RPC:s that are not in ietf-netconf.yang (and some others) that clixon implements. In those cases the only callbacks are the ones the application registers.
2. If you register an RPC that is in eg ietf-netconf.yang, you can
 a) return an error (either before/after) that will override the original.
 b) register _after_ and modify the original reply
 c) register _before_ and return a reply that takes precedence over the original. (in this case the backend will actually return two replies but the clients will only look at the first.

**Comment by olofhagsand** [comment]
But what I tried to formulate earlier is that if it is a common usecase to modify or preempt the system-defined RPC we could add some API to do exactly that, ie "removing" an existing callback, or something else?
When the callback API was designed the usecase of "double" callbacks was not really in scope, but it seems a useful one.

**Comment by Topitzik** [comment]
support for returning both errors and replies from user registered RPCs which are also system-defined, can be used to preempt RPCs by implementing an RPC that always failed. although such capability can never hurt.
In our case we wanted to only handle specific parameter cases, but were ready to preempt the RPC completely if it was not possible.
in the current implementation, I find (c) dangerous, the reply with precedence will also hide an error (which I assume should always have precedence by definition)

**Comment by olofhagsand** [comment]
Is it correct that your usecase is covered by (2b)?
If so, can we close this ticket?

**Comment by Topitzik** [comment]
yes, we can!
